mc-sema/cfgToLLVM/x86Instrs.cpp

```diff
@@ -169,12 +169,144 @@ InstTransResult liftInstrImpl(InstPtr ip, BasicBlock *&block, NativeBlockPtr nb,
     }
   } else {
     // Instruction translation not defined.
-    errs() << "Unsupported opcode " << opcode << "!";
+    errs() << "Unsupported opcode " << opcode << " ("
+      << IP->getOpcodeName(opcode) << "): ";
     IP->printInst( &inst, errs(), to_string<VA>(ip->get_loc(), hex));
-    // Print out the unhandled opcode.
-    errs() << to_string<VA>(ip->get_loc(), hex) << " ";
-    errs() << strOut.str() << "\n";
-    errs() << inst.getOpcode() << "\n";
+    errs() << "\n";
+
+    // In the case that we can't find the opcode, try building it out with
+    // inline assembly calls in LLVM instead. If at any point we don't know
+    // how to build the assembly, give up on trying to do so.
+    const MCRegisterInfo *MRI = natM->getMCContext()->getRegisterInfo();
+    const MCInstrInfo *MII = natM->getTarget()->createMCInstrInfo();
+
+    // Build the inline asm and the constraint string.
+    string inlineAsm = "", constraints = "";
+
+    // Use the printer to build the ASM string. We'll need to substitute
+    // $ for $$, though.
+    {
+      IP->printInst(&inst, strOut, "");
+      for (char c : strOut.str()) {
+        if (c == '$')
+          inlineAsm += "$$";
+        else
+          inlineAsm += c;
+      }
+    }
+
+    // Find all of the involved registers and build the LLVM
+    unsigned argnum = 0;
+    SmallVector<Value *, 3> operands;
+    SmallVector<Type *, 3> resultTypes;
+    for (unsigned i = 0; i < inst.getNumOperands(); i++) {
+      MCOperand &op = inst.getOperand(i);
+      if (op.isReg()) {
+        if (argnum++ > 0) constraints += ",";
+        if (i < MII->get(opcode).getNumDefs()) {
+          constraints += "=";
+
+          // Find the size of the register. Surprisingly, this is how you
```

Comment:
LLVM's MC is a bit annoying insofar as it has incomplete register dependency info. This shouldn't be too big of an issue though, given that the likely missing instructions are probably SSE or AVX ones.